src/analysis.zig

```diff
@@ -401,8 +361,89 @@ pub fn nodeToString(tree: *ast.Tree, node: *ast.Node) ?[]const u8 {
         },
         else => {
             std.debug.warn("INVALID: {}\n", .{node.id});
-        }
+        },
     }
-    
+
     return null;
 }
+
+pub const CompletionContext = struct {
+    pub const Id = enum {
+        builtin,
+        var_access,
+        field_access,
+        enum_literal,
+        container_field,
+        empty,
+        none,
+    };
+
+    id: Id,
+    node: ?*ast.Node = null,
+
+    fn toId(self: CompletionContext, id: Id) CompletionContext {
+        return .{ .id = id, .node = self.node };
+    }
+};
+
+fn nodeContainsSourceIndex(tree: *ast.Tree, node: *ast.Node, source_index: usize) bool {
+    const first_token = tree.tokens.at(node.firstToken());
+    const last_token = tree.tokens.at(node.lastToken());
+    return source_index >= first_token.start and source_index <= last_token.end;
+}
+
+fn visitNodesAndFindCompletionContext(tree: *ast.Tree, node: *ast.Node, source_index: usize) CompletionContext {
+    switch (node.id) {
+        .Identifier => {
+            var cc = CompletionContext{ .id = .var_access, .node = node };
+
+            if (nodeToString(tree, node).?[0] == '@') {
+                cc.id = .builtin;
+            }
+            return cc;
+        },
+        .InfixOp => {
+            const infix_op = node.cast(ast.Node.InfixOp).?;
+            if (infix_op.op == .Period and nodeContainsSourceIndex(tree, infix_op.rhs, source_index)) {
+                return .{ .id = .field_access, .node = node };
+            }
+        },
+        .BuiltinCall => return .{ .id = .builtin, .node = node },
+        .EnumLiteral => return .{ .id = .enum_literal, .node = node },
+        else => {},
+    }
+
+    var node_idx: usize = 0;
+    while (node.iterate(node_idx)) |child| : (node_idx += 1) {
+        if (!nodeContainsSourceIndex(tree, child, source_index)) continue;
+        return visitNodesAndFindCompletionContext(tree, child, source_index);
+    }
+
+    return switch (node.id) {
+        .Block, .Root => .{ .id = .empty },
+        .ContainerField => .{ .id = .container_field },
+        else => .{ .id = .none, .node = node },
+    };
+}
+
+pub fn completionContext(tree: *ast.Tree, source_index: usize) CompletionContext {
+    if (source_index == 0) return .{ .id = .empty };
+    const completion_context = visitNodesAndFindCompletionContext(tree, &tree.root_node.base, source_index);
+
+    // TODO: Can we do this better?
+    // Text like `someExpr().` will not get parsed to InfixOp.
+    // As a workaround, we check if the position corresponds exactly to a
+    // period token and adjust the context.
+    var token_idx: ast.TokenIndex = 0;
+    while (tree.tokens.at(token_idx).end <= source_index and token_idx < tree.tokens.len) : (token_idx += 1) {}
+    const tok_before_cursor = tree.tokens.at(token_idx);
+
+    const is_dot = tok_before_cursor.id == .Period and tok_before_cursor.end - 1 == source_index;
```

Comment:
You should also check that `tok_before_cursor.end - 1 != .Period`.
Currently this provides completions `foo[..<cursor here>)]`.

Reply:
This turned out to be a separate issue, see https://github.com/zigtools/zls/pull/41/commits/687addb49e925d2a1dcdd7e28086c37f709df0e2#diff-d2841187941f93280c97ecaafd02aeb4R423-R427